File: codeql/fixtures/csharp/redacted/ClaimService.cs
// Expected findings: 0.
//
// The same six flows as the violating fixture, each routed through the sanctioned
// redaction helper. This is the fixture that protects teams who did the right
// thing: if the barrier ever stops being recognised, this fails loudly here
// rather than quietly in fifty pull requests.
namespace Fixture.Redacted;

using Fixture.Support;

public sealed class ClaimService
{
    private readonly ILogger _logger;

    public ClaimService(ILogger logger) => _logger = logger;

    public void StructuredArgument(Customer customer, string claimNumber)
    {
        _logger.LogInformation(
            "Claim {ClaimNumber} for policyholder {Ahv}",
            claimNumber,
            PiiRedaction.MaskAhv(customer.AhvNumber));
    }

    public void Interpolated(Customer customer)
    {
        _logger.LogInformation($"Resolving policyholder {PiiRedaction.MaskAhv(customer.AhvNumber)}");
    }

    public void DateOfBirth(Customer customer)
    {
        _logger.LogWarning("Age check failed for {Dob}", PiiRedaction.MaskDateOfBirth(customer.DateOfBirth));
    }

    public void ViaLocal(Customer customer)
    {
        var ahvNumber = PiiRedaction.MaskAhv(customer.AhvNumber);
        _logger.LogError("Lookup failed for {Ahv}", ahvNumber);
    }

    public void PaymentFailure(Customer customer, string claimNumber)
    {
        _logger.LogError(
            "Payout for {ClaimNumber} failed to account {Iban}",
            claimNumber,
            PiiRedaction.MaskExceptSuffix(customer.Iban, 4));
    }

    public void ViaToString(Customer customer)
    {
        _logger.LogInformation("DOB {Dob}", PiiRedaction.MaskDateOfBirth(customer.DateOfBirth));
    }
}

File: codeql/fixtures/java/redacted/ClaimService.java
// Expected findings: 0.
//
// The violating flows routed through every sanitizer name family Pii.qll accepts:
// redact, mask, anonymi, pseudonymi, hash and tokeni.
import java.util.logging.Logger;

final class ClaimService {
    private final org.slf4j.Logger log;
    private final Logger auditLog = Logger.getLogger(ClaimService.class.getName());

    ClaimService(org.slf4j.Logger log) {
        this.log = log;
    }

    void structuredArgument(Customer customer, String claimNumber) {
        log.info(
                "Claim {} for policyholder {}",
                claimNumber,
                PiiMasking.maskAhv(customer.getAhvNumber()));
    }

    void concatenatedMessage(Customer customer) {
        log.info("Resolving policyholder " + PiiMasking.redact(customer.getAhvNumber()));
    }

    void viaLocal(Customer customer) {
        String safeValue = PiiMasking.anonymise(customer.getAhvNumber());
        log.error("Lookup failed for {}", safeValue);
    }

    void paymentFailure(Customer customer, String claimNumber) {
        log.error(
                "Payout for claim {} failed to account {}",
                claimNumber,
                PiiMasking.hash(customer.iban));
    }

    void viaToString(Customer customer) {
        log.warn("DOB {}", PiiMasking.maskDateOfBirth(customer.getDateOfBirth()));
    }

    void fieldAccess(Customer customer) {
        log.info("Manual review for {}", PiiMasking.pseudonymise(customer.policyholderName));
    }

    void javaUtilLogging(Customer customer) {
        auditLog.warning("Policyholder email " + PiiMasking.tokenise(customer.getEmailAddress()));
    }

    void consoleFallback(Customer customer) {
        System.err.printf("Policyholder %s%n", PiiMasking.maskAhv(customer.getAhvNumber()));
    }
}